src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -2171,26 +2266,45 @@ static const uint32_t *stream_normalize_pl (char * __restrict data, uint32_t * _
         break;
       case LENGTH_CODE_NEXTINT:
         /* NEXTINT */
-        if (!read_and_normalize_uint32 (&msz, data, off, size, bswap))
+        if (!read_and_normalize_uint32 (&msz, data, off, size1, bswap))
           return NULL;
         break;
       case LENGTH_CODE_ALSO_NEXTINT: case LENGTH_CODE_ALSO_NEXTINT4: case LENGTH_CODE_ALSO_NEXTINT8:
         /* length is part of serialized data */
-        if (!peek_and_normalize_uint32 (&msz, data, off, size, bswap))
+        if (!peek_and_normalize_uint32 (&msz, data, off, size1, bswap))
           return NULL;
         if (lc > LENGTH_CODE_ALSO_NEXTINT)
-          msz <<= (lc - 4);
+        {
+          uint32_t shift = lc - 4;
+          if (msz > UINT32_MAX >> shift)
+            return NULL;
+          msz <<= shift;
+        }
+        /* length embedded in member does not include it's own 4 bytes, we need to be able
+           to add those 4; technically perhaps this would be valid CDR but if so, we don't
+           support it */
+        if (msz > UINT32_MAX - 4)
+          return NULL;
+        else
+          msz += 4;
         break;
       default:
         abort ();
         break;
     }
-
-    if (!dds_stream_normalize_pl_member (data, m_id, off, size, bswap, xcdr_version, ops))
+    // reject if fewer than msz bytes remain in declared size of the parameter list
+    if (msz > size1 - *off)
+      return NULL;
+    // don't allow member values that exceed its declared size
+    const uint32_t size2 = *off + msz;
+    switch (dds_stream_normalize_pl_member (data, m_id, off, size2, bswap, xcdr_version, ops))
     {
-      *off += msz;
-      if (lc >= LENGTH_CODE_ALSO_NEXTINT)
-        *off += 4; /* length embedded in member does not include it's own 4 bytes */
+      case NPMR_NOT_FOUND:
+      case NPMR_FOUND:
+        *off = size2;
```

Comment:
I think we can add a check here that `*off` must be equal to `size2`: in case a member is found the data should have the length we expect for this member type?

Reply:
Agree.